Content.Server/StationEvents/Events/SecondLifeRule.cs

```diff
@@ -0,0 +1,181 @@
+                {
+                    ghostInfo.Rules = component.Rules;
+                    ghostInfo.Description = component.Description;
+
+                    // Only need to explicitly call UpdateAllEui if we only update rules and desc without
+                    //   adding or removing roles after.
+                    rolesDirty = true;
+                }
+            }
+            else
+            {
+                // This role requires registration. Perhaps the event just started or perhaps someone
+                //   added / increased limit on a role with VV
+
+                if (PrototypeManager.TryIndex<JobPrototype>(roleKvp.Key, out var job))
+                {
+                    var desc = (job.Description != null)
+                        ? $"{Loc.GetString(job.Description)} : {Loc.GetString(component.Description)}"
+                        : Loc.GetString(component.Description);
+
+                    var info = new GhostRoleInfo
+                    {
+                        Name = Loc.GetString(job.Name),
+                        Description = desc,
+                        Rules = Loc.GetString(component.Rules),
+                        Owner = uid, // All the Roles point back to the SecondLifeRuleComponent's entity.
+                        UserId = roleKvp.Key,
+                    };
+                    _ghostRole.RegisterGhostRole(ref info);
+                    component.RoleInfo[info.UserId] = info;
+                }
+            }
+        }
+
+        List<string>? toRemove = null;
+        foreach (var infoKvp in component.RoleInfo)
+        {
+            // Check this role is still within the Roles dictionary (incase VV deleted it)
+            if (!component.Roles.ContainsKey(infoKvp.Key))
+            {
+                _ghostRole.UnregisterGhostRole(infoKvp.Value);
+                if (toRemove == null)
+                    toRemove = new();
+
+                toRemove.Add(infoKvp.Key);
+            }
+        }
+
+        if (toRemove != null)
+        {
+            // Do the removal outside the component.RuleInfo iteration
+            foreach (var rolename in toRemove)
+            {
+                component.RoleInfo.Remove(rolename);
+            }
+        }
+
+        if (rolesDirty)
+            _ghostRole.UpdateAllEui();
+    }
+
+    private void DeregisterAllRoles(EntityUid uid, SecondLifeRuleComponent component)
+    {
+        foreach (var infoKvp in component.RoleInfo)
+        {
+            // Check this role is still within the Roles dictionary (incase VV deleted it)
+            _ghostRole.UnregisterGhostRole(infoKvp.Value);
+        }
+        component.RoleInfo.Clear();
+    }
+
+    private void OnTakeGhostRole(EntityUid uid, SecondLifeRuleComponent component, ref TakeGhostRoleEvent args)
+    {
+        args.TookRole = false;
+        if (!component.Roles.TryGetValue(args.UserId, out var limit))
+            return;
```

Comment:
so couple thoughts on who should be allowed to take these ghost roles:
- does this take into account people who observe from the beginning of a round? actually-- since it doesnt update total number of players spawned if its from a secondlife, presumably these people wont count to the total even though they were never in the round to begin with. imo they should be disallowed altogether (if they wanted to play, they should have joined the game regularly)
- i dont think you should be able to die and immediately take a second life ghost role if one exists. there should be some cooldown (10-15 min?) that you have to be dead for before the game allows you to. too little and depending on how common these are people might care less about dying since they know they can respawn immediately
- people that suicide should be forbidden from taking second life ghost roles--if you werent willing to live out your first life you dont get a second one imo. this is metagamable by just dying through other means but its better than nothing. there should already be an ~easy way to check for this (cloning does)?

Reply:
Okay, these are fair points. We'd need to track "eligible ghosts" who are allowed to take on good ghost roles and who meet this criteria. To be clear:

1. Joined round as player (NOT Observer)
2. Have not used suicide command
3. Have been dead for 10-15 minutes

Are you sure about 10-15? That's a long time. Perhaps 5 minutes? Even that seems like a long time if you're sitting around. 

Hmmm quite a bit more tracking code required for this.